### PR TITLE
[Qualityathon]: test - static analysis - format

### PR DESCRIPTION
**Pull request by fhboswell**
# Description
## One Line Summary
Adding testing static analysis and code formatting for the Qualityathon

### Motivation
Testing, static analysis and code formatting as part of git actions and precommit hooks will improve the quality of the code that gets pulled into the project.

### Scope
Testing will be done with Jest
Static Analysis will be done with eslint
Formatting will be done with prettier

# Affected code checklist
   - [x] Testing, static analysis and code formatting as part of git actions and precommit hooks will improve the quality of the code that gets pulled into the project.

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [x] All automated tests pass, or I explained why that is not possible
   - [ ] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [ ] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [ ] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/onesignal/react-native-onesignal/1426)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+on: push
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Install modules
+      run: yarn
+    - name: Run tests
+      run: yarn test
```
Comment (rgomezp) [inline .github/workflows/ci.yml]: newline

File: __test__/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+describe("Test for initial Jest setup.", () => {
+    test('two plus two is four', () => {
+        expect(2 + 2).toBe(4);
+      });
+});
```
Comment (rgomezp) [inline __test__/index.test.ts]: newline

File: babel.config.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  presets: ['module:metro-react-native-babel-preset'],
+};
```
Comment (rgomezp) [inline babel.config.js:3]: why was this added?

File: coverage/clover.xml [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+      </file>
+      <file name="Outcomes.ts" path="/Users/henryboswell/Desktop/react-native-onesignal/src/models/Outcomes.ts">
+        <metrics statements="0" coveredstatements="0" conditionals="0" coveredconditionals="0" methods="0" coveredmethods="0"/>
+      </file>
+      <file name="Subscription.ts" path="/Users/henryboswell/Desktop/react-native-onesignal/src/models/Subscription.ts">
+        <metrics statements="0" coveredstatements="0" conditionals="0" coveredconditionals="0" methods="0" coveredmethods="0"/>
+      </file>
+    </package>
+  </project>
+</coverage>
```
Comment (rgomezp) [inline coverage/clover.xml]: what does this file do? add a bit more details in individual commits

File: coverage/coverage-final.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
...start":{"line":72,"column":12},"end":{"line":72,"column":63}},"30":{"start":{"line":73,"column":12},"end":{"line":73,"column":63}},"31":{"start":{"line":74,"column":12},"end":{"line":74,"column":67}},"32":{"start":{"line":75,"column":12},"end":{"line":75,"column":63}},"33":{"start":{"line":76,"column":12},"end":{"line":76,"column":69}}},"fnMap":{"0":{"name":"(anonymous_0)","decl":{"start":{"line":39,"column":4},"end":{"line":39,"column":5}},"loc":{"start":{"line":39,"column":47},"end":{"line":78,"column":5}},"line":39}},"branchMap":{"0":{"loc":{"start":{"line":49,"column":8},"end":{"line":62,"column":9}},"type":"if","locations":[{"start":{"line":49,"column":8},"end":{"line":62,"column":9}},{"start":{},"end":{}}],"line":49},"1":{"loc":{"start":{"line":64,"column":8},"end":{"line":77,"column":9}},"type":"if","locations":[{"start":{"line":64,"column":8},"end":{"line":77,"column":9}},{"start":{},"end":{}}],"line":64}},"s":{"0":0,"1":0,"2":0,"3":0,"4":0,"5":0,"6":0,"7":0,"8":0,"9":0,"10":0,"11":0,"12":0,"13":0,"14":0,"15":0,"16":0,"17":0,"18":0,"19":0,"20":0,"21":0,"22":0,"23":0,"24":0,"25":0,"26":0,"27":0,"28":0,"29":0,"30":0,"31":0,"32":0,"33":0},"f":{"0":0},"b":{"0":[0,0],"1":[0,0]}}
```
Comment (rgomezp) [inline coverage/coverage-final.json]: what does this file do? the path here is machine-specific

File: coverage/lcov-report/OSNotification.ts.html [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+<span class="cline-any cline-no">&nbsp;</span>
+<span class="cline-any cline-no">&nbsp;</span>
+<span class="cline-any cline-no">&nbsp;</span>
+<span class="cline-any cline-no">&nbsp;</span>
+<span class="cline-any cline-neutral">&nbsp;</span>
+<span class="cline-any cline-neutral">&nbsp;</span>
+<span class="cline-any cline-neutral">&nbsp;</span>
+<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import { Platform } from 'react-native';
+&nbsp;
+export default class OSNotification {
```
Comment (rgomezp) [inline coverage/lcov-report/OSNotification.ts.html]: is this supposed to be there?

File: __test__/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+
+describe("Test for initial jest setup.", () => {
+    test('two plus two is four', () => {
+        expect(2 + 2).toBe(4);
+      });
+});
```
Comment (rgomezp) [inline __test__/index.test.ts]: newline

File: jest.config.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    collectCoverageFrom: ["src/**/*"],
+    moduleFileExtensions: [
+        "ts",
+        "tsx",
+        "js",
+        "jsx",
+        "json",
+        "node"
+    ]
+}
```
Comment (rgomezp) [inline jest.config.js]: newline

File: src/events/EventManager.ts [hunk 9]
```diff
@@ -1,133 +1,143 @@
-import { EmitterSubscription, NativeEventEmitter, NativeModule } from 'react-native';
+import {
+  EmitterSubscription,
+  NativeEventEmitter,
+  NativeModule,
+} from 'react-native';
 import NotificationReceivedEvent from './NotificationReceivedEvent';
-import { isMultipleInstancesPossible } from '../helpers';
+import {isMultipleInstancesPossible} from '../helpers';
```
Comment (rgomezp) [inline src/events/EventManager.ts]: there should be spaces between bracket and name. probably a rule that needs to be added.

for consistency with other SDKs

**Review by rgomezp — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jmadler) [reply .github/workflows/ci.yml]: 👍

File: __test__/index.test.ts (hunk 2 above)
Comment (jmadler) [reply __test__/index.test.ts]: 👍

File: coverage/coverage-final.json (hunk 5 above)
Comment (jmadler) [reply coverage/coverage-final.json]: ^

File: coverage/lcov-report/OSNotification.ts.html (hunk 6 above)
Comment (jmadler) [reply coverage/lcov-report/OSNotification.ts.html]: ^

File: coverage/clover.xml (hunk 4 above)
Comment (jmadler) [reply coverage/clover.xml]: The coverage directory was gitignored and doesn't exist on my checkout.  ~I'll purge it from the branch.~ Looks like it's marked as a part of an outdated commit, it should be gone.

File: __test__/index.test.ts (hunk 7 above)
Comment (jmadler) [reply __test__/index.test.ts]: 👍

File: jest.config.js (hunk 8 above)
Comment (jmadler) [reply jest.config.js]: 👍

File: babel.config.js (hunk 3 above)
Comment (jmadler) [reply babel.config.js:3]: I believe it was added by a [new dependency](https://github.com/OneSignal/react-native-onesignal/pull/1426/commits/0e7be2a1e5852424b1af1ce4b2cad8971b2bc832#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519).  @fhboswell might have more info

File: src/events/EventManager.ts (hunk 9 above)
Comment (jmadler) [reply src/events/EventManager.ts]: I've pushed a commit with bracketSpacing enabled.  Let me know what you think. cc @fhboswell

File: coverage/clover.xml (hunk 4 above)
Comment (fhboswell) [reply coverage/clover.xml]: the coverage directory is in the .gitignore

File: babel.config.js (hunk 3 above)
Comment (fhboswell) [reply babel.config.js:3]: This is required to convert typescript to javascript for jest to run. the example directory uses this as well. Open to alternatives.

**Review by rgomezp — APPROVED** [review]
